src/linux_mcp_server/tools/system_info.py

```diff
@@ -1,273 +1,453 @@
-
-            # CPU frequency
-            try:
-                cpu_freq = psutil.cpu_freq()
-                if cpu_freq:
-                    info.append(
-                        f"CPU Frequency: Current={cpu_freq.current:.2f}MHz, Min={cpu_freq.min:.2f}MHz, Max={cpu_freq.max:.2f}MHz"
-                    )
-            except Exception:
-                pass  # CPU frequency might not be available
-
-            # CPU usage per core
-            cpu_percent = psutil.cpu_percent(interval=0.1, percpu=True)
-            info.append("\nCPU Usage per Core:")
-            for i, percent in enumerate(cpu_percent):
-                info.append(f"  Core {i}: {percent}%")
-
-            # Overall CPU usage
-            overall_cpu = psutil.cpu_percent(interval=0.1)
-            info.append(f"\nOverall CPU Usage: {overall_cpu}%")
-
-            # Load average
-            load_avg = os.getloadavg()
-            info.append(f"\nLoad Average (1m, 5m, 15m): {load_avg[0]:.2f}, {load_avg[1]:.2f}, {load_avg[2]:.2f}")
-
-            # Try to get CPU model info from /proc/cpuinfo
-            try:
-                with open("/proc/cpuinfo") as f:
-                    for line in f:
-                        if line.startswith("model name"):
-                            cpu_model = line.split(":")[1].strip()
-                            info.insert(0, f"CPU Model: {cpu_model}")
-                            break
-            except Exception:
-                pass
-
-        return "\n".join(info)
-    except Exception as e:
-        return f"Error gathering CPU information: {str(e)}"
+@mcp.tool(
+    title="Get CPU information",
+    description="Get CPU information.",
+    annotations=ToolAnnotations(readOnlyHint=True),
+)
+@log_tool_call
+async def get_cpu_information(
+    host: Host | None = None,
+    username: Username | None = None,
+    fields: t.Annotated[
+        list[str],
+        Field(
+            description="List of fields to include. If None, all fields are included. Valid fields: cpu_model, logical_cores, physical_sockets, physical_cores_per_processor, frequency_mhz, load_average, cpu_usage"
+        ),
+    ]
+    | None = None,
+) -> CPUInfo:
+    """Get CPU information."""
+
+    # Collect phase: Execute all unique commands (note: /proc/cpuinfo only read once!)
+    commands = [
+        ["cat", "/proc/cpuinfo"],
+        ["cat", "/proc/loadavg"],
+        ["top", "-bn1"],
+    ]
+    raw_outputs = await collect_command_outputs(commands, host=host, username=username)
+
+    # Parse phase: Convert raw outputs to structured data
+    parsed_data = parse_cpu_information(raw_outputs)
+
+    # Filter phase: Apply field filters and build result
+    filtered_data = filter_fields(parsed_data, fields)
+
+    # Build CPUInfo object from filtered data
+    load_avg_data = filtered_data.get("load_average")
+    load_average = LoadAverage(**load_avg_data) if load_avg_data else None
```

Comment:
Instead of assigning `None` in case of not having data for `load_avg_data`, I think we should still assign a `LoadAverage` instance, but with default values. 

This ensures that we maintain a structured output, regardless if we have data to it or not, avoiding a case where a later read to CPUInfo.load_average might be None. 

```suggestion
    load_average = LoadAverage(**load_avg_data) if load_avg_data else LoadAverage()
```